### PR TITLE
Decouple subdivision targeting codes from Rewards.

### PR DESCRIPTION
**Pull request by aseren**
Now subdivision targeting codes are defined in one file.

Resolves https://github.com/brave/brave-browser/issues/25249

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 1]
```diff
@@ -1197,11 +1195,22 @@ void RewardsDOMHandler::GetAdsData(const base::Value::List& args) {
...
+  ads_data.Set("shouldAllowAdsSubdivisionTargeting",
                ads_service_->ShouldAllowSubdivisionTargeting());
   ads_data.Set("adsUIEnabled", true);
-  ads_data.Set(kNeedsBrowserUpgradeToServeAds,
+  ads_data.Set("needsBrowserUpgradeToServeAds",
                ads_service_->NeedsBrowserUpgradeToServeAds());
+
+  base::Value::List subdivisions;
+  const auto subdivision_infos = ads::GetDefaultSupportedSubdivisions();
+  for (const auto& info : subdivision_infos) {
```
Comment (tmancey) [inline browser/ui/webui/brave_rewards_page_ui.cc:1206]: nit: `for (const auto& subdivision : supported_subdivisions) {`

File: vendor/bat-native-ads/include/bat/ads/default_supported_subdivisions.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_ADS_INCLUDE_BAT_ADS_DEFAULT_SUPPORTED_SUBDIVISIONS_H_
+#define BRAVE_VENDOR_BAT_NATIVE_ADS_INCLUDE_BAT_ADS_DEFAULT_SUPPORTED_SUBDIVISIONS_H_
+
+#include "base/containers/flat_map.h"
+#include "base/strings/string_piece.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/include/bat/ads/default_supported_subdivisions.h:10]: nit: `#include "base/strings/string_piece_forward.h"`

**Review by zenparsing — APPROVED** [review]
Rewards front-end changes 👍

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivisions.cc [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+bool IsSupportedSubdivisionCode(const std::string& country_code,
+                                const std::string& subdivision_code) {
+  const auto iter = GetSupportedSubdivisionInfos().find(country_code);
+  if (iter == GetSupportedSubdivisionInfos().cend()) {
+    return false;
+  }
+
+  const auto& subdivisions = iter->second;
+  return subdivisions.find(subdivision_code) != subdivisions.cend();
+}
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivisions.cc:79]: `IsSupported...` could be moved to a subdivision_code_util and then his file could be supported_subdivision_codes

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivisions.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_ADS_SRC_BAT_ADS_INTERNAL_GEOGRAPHIC_SUBDIVISION_SUPPORTED_SUBDIVISIONS_H_
+#define BRAVE_VENDOR_BAT_NATIVE_ADS_SRC_BAT_ADS_INTERNAL_GEOGRAPHIC_SUBDIVISION_SUPPORTED_SUBDIVISIONS_H_
+
+#include <string>
+
+#include "base/containers/flat_map.h"
+#include "base/strings/string_piece.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivisions.h:12]: see previous comment about `forward`

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/subdivision_targeting.h [hunk 5]
```diff
@@ -9,6 +9,7 @@
 #include <string>
 
 #include "absl/types/optional.h"
+#include "base/strings/string_piece.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/subdivision_targeting.h:12]: is this used?

File: vendor/bat-native-ads/src/bat/ads/default_supported_subdivisions.cc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/default_supported_subdivisions.h"
+
+#include "base/no_destructor.h"
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/default_supported_subdivisions.cc:8]: not used

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                {"US-NM", "New Mexico"},     {"US-NY", "New York"},
+                {"US-NC", "North Carolina"}, {"US-ND", "North Dakota"},
+                {"US-OH", "Ohio"},           {"US-OK", "Oklahoma"},
+                {"US-OR", "Oregon"},         {"US-PA", "Pennsylvania"},
+                {"US-RI", "Rhode Island"},   {"US-SC", "South Carolina"},
+                {"US-SD", "South Dakota"},   {"US-TN", "Tennessee"},
+                {"US-TX", "Texas"},          {"US-UT", "Utah"},
+                {"US-VT", "Vermont"},        {"US-VA", "Virginia"},
+                {"US-WA", "Washington"},     {"US-WV", "West Virginia"},
+                {"US-WI", "Wisconsin"},      {"US-WY", "Wyoming"},
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc:46]: nit: remove last comma

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                // Canada
+                {"CA-AB", "Alberta"},
+                {"CA-BC", "British Columbia"},
+                {"CA-MB", "Manitoba"},
+                {"CA-NB", "New Brunswick"},
+                {"CA-NS", "Nova Scotia"},
+                {"CA-ON", "Ontario"},
+                {"CA-QC", "Quebec"},
+                {"CA-SK", "Saskatchewan"},
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc:58]: nit: remove last comma

File: vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            }}});
+
+  return *supported_subdivision_codes;
+}
+
+bool IsSupportedSubdivisionCode(const std::string& country_code,
+                                const std::string& subdivision_code) {
+  const auto it = GetSupportedSubdivisionCodes().find(country_code);
+  if (it == GetSupportedSubdivisionCodes().cend()) {
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/geographic/subdivision/supported_subdivision_codes.cc:67]: very small nit: for consistency in ads could we use `iter`?

File: vendor/bat-native-ads/src/bat/ads/supported_subdivisions.cc [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+#include "bat/ads/internal/geographic/subdivision/supported_subdivision_codes.h"
+#include "brave/components/l10n/common/locale_util.h"
+
+namespace ads {
+
+SupportedSubdivisions GetSupportedSubdivisions() {
+  const std::string country_code = brave_l10n::GetDefaultISOCountryCodeString();
+
+  const auto& subdivision_codes = geographic::GetSupportedSubdivisionCodes();
+  const auto it = subdivision_codes.find(country_code);
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/supported_subdivisions.cc:18]: See small nit above

File: vendor/bat-native-ads/src/bat/ads/internal/base/locale/subdivision_code_util.cc [hunk 11]
```diff
@@ -9,9 +9,20 @@
 
 #include "base/check_op.h"
 #include "base/strings/string_split.h"
+#include "bat/ads/internal/geographic/subdivision/supported_subdivision_codes.h"
+#include "brave/components/l10n/common/locale_util.h"
 
 namespace ads::locale {
 
+bool IsSupportedLocaleForSubdivisionTargeting(const std::string& locale) {
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/base/locale/subdivision_code_util.cc:17]: nit: maybe change to IsSupportedCountryCodeForSubdivisionTargeting and pass in `brave_l10n::GetISOCountryCode(locale)`

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by goodov — APPROVED** [review]
lgtm with few nits

**Pull request merged**